mixnet/node/src/client_notifier.rs

```diff
@@ -10,30 +11,41 @@ pub struct ClientNotifier {}
 
 impl ClientNotifier {
     pub async fn run(
-        mut listen_address: SocketAddr,
+        listen_address: Arc<Atomic<SocketAddr>>,
         mut rx: mpsc::Receiver<Body>,
+        shutdown_rx: async_channel::Receiver<()>,
     ) -> Result<(), Box<dyn Error>> {
-        let listener = TcpListener::bind(listen_address).await?;
+        let mut addr = listen_address.load(Ordering::Acquire);
+        let listener = TcpListener::bind(addr).await?;
 
         // update the port if the port is assigned automatically by the system
-        if listen_address.port() == 0 {
-            listen_address.set_port(listener.local_addr().unwrap().port());
+        if addr.port() == 0 {
+            addr.set_port(listener.local_addr().unwrap().port());
+            listen_address.store(addr, Ordering::Release);
         }
 
-        tracing::info!("Listening mixnet client connections: {listen_address}");
+        tracing::info!("Listening mixnet client connections: {addr}");
```

Comment:
It'd be better to separate the port problem with this graceful shutdown PR, if possible.

I guess we can solve the port problem without using `Arc` because the port is initialized at once and never changes. For example, we could probably make a `MixnetNode::run()` function which calls `TcpListener::bind()`s for all necessary ports, such as `listen_address` and `client_listen_address`. (we can probably make a `ClientNotifier::new()` function as well that does only `bind()`).
In this way, I guess we don't need `Arc` even though I haven't tried it yet.

Reply:
Ah, I got you, I will change the code.

Reply:
It seems this PR still contains two things: auto-port and graceful-shutdown. Could we move auto-port related fixes to the PR #327, so that this PR contains only gradeful-shutdown stuff?

Reply:
Yes, I forgot to remove the changes in the test file.